chore(cohorts): Remove the old way of calculating cohorts

haacked opened the pull request:
We now use hogql for calculating cohorts. This was controlled by the `enable_hogql_cohort_calculation` flag which has been rolled out to 100% for a while and we're not going back.

> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

I see 💀 code. I want to get rid of it.

## Changes

Removed the branch where `enable_hogql_cohort_calculation` is `false`.

## Did you write or update any docs for this change?

Nope

## How did you test this code?

Manually.
Updated unit tests.

haacked committed:
Remove the old way of calculating cohorts

We now use hogql for calculating cohorts. The flag has been 100% for a while and we're not going back.

haacked commented on posthog/api/test/test_cohort.py:
```diff
@@ -263,7 +263,7 @@ def test_action_persons_on_events(self, patch_sync_execute, patch_calculate_coho
             )
 
-            self.assertIn(f"mat_pp_team_id", insert_statements[0])
+            self.assertIn(f"JSONExtractRaw(person.properties, 'team_id')", insert_statements[0])
```
Hmmm, this seems like a bug to me the more I learn about ClickHouse. We should probably fix it.

haacked committed:
Fix these tests

The property values we create are strings, so it makes sense that the filter should also be strings.

haacked commented on ee/clickhouse/queries/test/test_person_query.py:
```diff
@@ -91,7 +91,7 @@ def test_person_query_with_multiple_cohorts(testdata, team, snapshot):
                         "values": [
                             {"key": "group", "value": "none", "type": "person"},
-                            {"key": "group", "value": [1, 2, 3], "type": "person"},
+                            {"key": "group", "value": ["1", "2", "3"], "type": "person"},
```
Prior to this change, the HogQL query was generating this as a WHERE clause:

```sql
WHERE in(persons.properties___group, tuple(1.0, 2.0, 3.0))
```

Which didn't match anything, causing the test to fail because the Cohort it created was empty. It appears HogQL is a bit more strict around this perhaps and the old code more forgiving.

By changing these to strings, it changes the query to:

```sql
in(persons.properties___group, tuple(%(hogql_val_13)s, %(hogql_val_14)s, %(hogql_val_15)s))
```

haacked commented on posthog/api/test/test_feature_flag.py:
```diff
@@ -6077,7 +6078,7 @@ def test_user_blast_radius_with_single_cohort(self):
                             "values": [
                                 {"key": "group", "value": "none", "type": "person"},
-                                {"key": "group", "value": [1, 2, 3], "type": "person"},
+                                {"key": "group", "value": ["1", "2", "3"], "type": "person"},
```
What's interesting is I expected this to fail because the filter values are strings and the group values are integers, but it passes.

In fact, if I make the filter values integers, then this fails. @aspicer do you have any insights into why that is?

haacked commented:
Removing the feature flag broke some tests. That means either the tests were wrong, or behavior is broken. I was able to fix some tests, but I'm still trying to understand the other breakages.

The one I'm trying to track down now is `test_create_exposure_cohort_for_experiment_with_custom_action_filters_exposure` which used to return two users but now returns one and it's not immediately apparent why. Here's the HogQL query generated for created the experimental cohort.

```sql
INSERT INTO cohortpeople
SELECT id, %(cohort_id)s as cohort_id, %(team_id)s as team_id, 1 AS sign, %(new_version)s AS version
FROM (
    (SELECT
    source.id AS id
FROM
    (SELECT
        actor_id AS actor_id,
        count() AS event_count,
        groupUniqArray(distinct_id) AS event_distinct_ids,
        actor_id AS id
    FROM
        (SELECT
            if(not(empty(e__override.distinct_id)), e__override.person_id, e.person_id) AS actor_id,
            toTimeZone(e.timestamp, %(hogql_val_0)s) AS timestamp,
            e.uuid AS uuid,
            e.distinct_id AS distinct_id
        FROM
            events AS e
            LEFT OUTER JOIN (SELECT
                argMax(person_distinct_id_overrides.person_id, person_distinct_id_overrides.version) AS person_id,
                person_distinct_id_overrides.distinct_id AS distinct_id
            FROM
                person_distinct_id_overrides
            WHERE
                equals(person_distinct_id_overrides.team_id, 1311)
            GROUP BY
                person_distinct_id_overrides.distinct_id
            HAVING
                ifNull(equals(argMax(person_distinct_id_overrides.is_deleted, person_distinct_id_overrides.version), 0), 0)
            SETTINGS optimize_aggregation_in_order=1) AS e__override ON equals(e.distinct_id, e__override.distinct_id)
        WHERE
            and(equals(e.team_id, 1311), and(ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_1)s), ''), 'null'), '^"|"$', ''), %(hogql_val_2)s), 0), in(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_3)s), ''), 'null'), '^"|"$', ''), tuple(%(hogql_val_4)s, %(hogql_val_5)s))), greaterOrEquals(timestamp, toDateTime64('2025-05-18 03:56:06.857755', 6, 'UTC')), lessOrEquals(timestamp, toDateTime64('2025-05-23 23:59:59.999999', 6, 'UTC')), or(and(equals(e.event, %(hogql_val_6)s), and(ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_7)s), ''), 'null'), '^"|"$', ''), %(hogql_val_8)s), 0), in(if(not(empty(e__override.distinct_id)), e__override.person_id, e.person_id), (SELECT
                                    cohortpeople.person_id AS person_id
                                FROM
                                    cohortpeople
                                WHERE
                                    and(equals(cohortpeople.team_id, 1311), equals(cohortpeople.cohort_id, 1571), equals(cohortpeople.version, 1)))))), and(equals(e.event, %(hogql_val_9)s), ifNull(greater(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_10)s), ''), 'null'), '^"|"$', ''), %(hogql_val_11)s), 0)), and(equals(e.event, %(hogql_val_12)s), ifNull(match(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_13)s), ''), 'null'), '^"|"$', ''), %(hogql_val_14)s), isNull(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, %(hogql_val_13)s), ''), 'null'), '^"|"$', '')) and isNull(%(hogql_val_14)s))))))
    GROUP BY
        actor_id) AS source
ORDER BY
    source.id ASC
LIMIT 1000000000
SETTINGS optimize_aggregation_in_order=1, join_algorithm='auto')
) as person
UNION ALL
SELECT person_id, cohort_id, team_id, -1, version
FROM cohortpeople
WHERE team_id = %(team_id)s AND cohort_id = %(cohort_id)s AND version < %(new_version)s AND sign = 1
SETTINGS optimize_aggregation_in_order = 1, join_algorithm = 'auto'
```

aspicer committed:
move

aspicer commented:
@haacked for that test `test_create_exposure_cohort_for_experiment_with_custom_action_filters_exposure`, I just had to move `cohort_extra.calculate_people_ch(pending_version=1)` after `flush_persons_and_events`

For a reason I did not dig into, in this HogQL generated clickhouse, the "in cohort" clause wasn't expanded into its constituent parts but left as an `person_id in (select ... from cohortpeople)` clause

Since we only recalculated the other cohort at the beginning of the function and not after we created all the people, it was empty, so nothing was matching.

aspicer committed:
property definitions

aspicer commented:
As for `test_person_cohorts` I had to add a property definition setting the numeric type as a numeric in order to avoid the "supertype" error

```
PropertyDefinition.objects.create(
            team=self.team, name="number", property_type="Numeric", type=PropertyDefinition.Type.PERSON
        )
```

HogQL is stricter about property definitions and does different logic based on the type. The old queries were somewhat untyped and allowed more looseness.

haacked commented:
Thanks @aspicer!

haacked committed:
Rename property to not be confusing

haacked committed:
Merge branch 'master' into haacked/remove-enable_hogql_cohort_calculation-feature-flag

haacked committed:
Fix the test

Cohort queries no longer looks at the `person_properties` column on the `events` table because those values get out of date. Instead, it looks directly at `person.properties`. So if we materialize a column, we need to do it on the `person` table, not `events`. After doing so, the query leverages the materialized column.

haacked committed:
Fix assert and add clarifying comment

haacked committed:
Merge branch 'master' into haacked/remove-enable_hogql_cohort_calculation-feature-flag

haacked committed:
Merge branch 'master' into haacked/remove-enable_hogql_cohort_calculation-feature-flag
